msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ConfidentialClientApplication.java

```diff
@@ -93,23 +98,58 @@ private void initClientAuthentication(IClientCredential clientCredential) {
         }
     }
 
+    /**
+     * Generates a JWT-formatted assertion string based on the provided client credential. Returns null in cases where
+     * the request for that credential type would not use a JWT assertion (e.g. client secret).
+     *
+     * @param clientCredential  The client credential to use for token acquisition.
+     * @return JWT-formatted assertion string
+     */
     String getAssertionString(IClientCredential clientCredential) {
         if (clientCredential instanceof ClientCertificate) {
-            boolean useSha1 = Authority.detectAuthorityType(this.authenticationAuthority.canonicalAuthorityUrl()) == AuthorityType.ADFS;
-
-            return JwtHelper.buildJwt(
-                    clientId(),
-                    clientCertificate,
-                    this.authenticationAuthority.selfSignedJwtAudience(),
-                    sendX5c,
-                    useSha1).assertion();
+            // Check if the current assertion is null or has expired, and if so create a new one
+            if (this.assertion == null || hasJwtExpired(this.assertion)) {
+                boolean useSha1 = Authority.detectAuthorityType(this.authenticationAuthority.canonicalAuthorityUrl()) == AuthorityType.ADFS;
+
+                this.assertion = JwtHelper.buildJwt(
```

Comment:
1. The assertion should not be cached. 
2. The assertion contains the issuer, which is the Entra token endpoint. The token endpoint contains the tenant. So you can't resolve the assertion at the Application level, only at the request level, where the user can override the tenant?